agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpClientTracer.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.auto.bootstrap.instrumentation.decorator;
+
+import static io.opentelemetry.context.ContextUtils.withScopedContext;
+import static io.opentelemetry.trace.TracingContextUtils.withSpan;
+
+import io.grpc.Context;
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.auto.config.Config;
+import io.opentelemetry.auto.instrumentation.api.MoreAttributes;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.DefaultSpan;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.TracingContextUtils;
+import io.opentelemetry.trace.attributes.SemanticAttributes;
+import java.net.URI;
+import java.net.URISyntaxException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class HttpClientTracer<REQUEST, RESPONSE> extends BaseTracer {
+
+  private static final Logger log = LoggerFactory.getLogger(HttpClientTracer.class);
+
+  public static final String DEFAULT_SPAN_NAME = "HTTP request";
+
+  protected static final String USER_AGENT = "User-Agent";
+
+  protected abstract String method(REQUEST request);
+
+  protected abstract URI url(REQUEST request) throws URISyntaxException;
+
+  protected abstract Integer status(RESPONSE response);
+
+  protected abstract String requestHeader(REQUEST request, String name);
+
+  protected abstract String responseHeader(RESPONSE response, String name);
+
+  protected HttpTextFormat.Setter<REQUEST> getSetter() {
+    throw new UnsupportedOperationException();
+  }
+
+  public Span startSpan(REQUEST request) {
+    return startSpan(request, spanNameForRequest(request));
+  }
+
+  public Scope startScope(Span span, REQUEST request) {
+    Context context = withSpan(span, Context.current());
+    OpenTelemetry.getPropagators().getHttpTextFormat().inject(context, request, getSetter());
+    return withScopedContext(context);
+  }
+
+  // TODO should end methods remove SPAN attribute from request as well?
```

Comment:
This comment is not relevant here. You don't attach span to requests.

Reply:
Some of the Setters are not <REQUEST> type, it failed to work for request.builder or HttpHeaders.  There are quite a few.  For those cases, getSetter() will simply return null.  Initially i had it abstract, but then later realized it's better to provide a default implementation so that it's not a required override method in the sub-tracers.

Reply:
Sorry, @heyams, I don't understand how your comment is relevant to this comment on `end` method?

Reply:
to your first comment: some of the Setters are not REQUEST type, it failed to work for request.builder or HttpHeaders. There are quite a few. For those cases, getSetter() will simply return null. Initially i had it abstract, but then later realized it's better to provide a default implementation so that it's not a requires override method in the sub-tracers.